src/coprocessor/mod.rs

```diff
@@ -53,59 +54,112 @@ trait RequestHandler: Send {
         // Do nothing by default
     }
 
-    fn into_boxed(self) -> Box<RequestHandler>
+    fn into_boxed(self) -> Box<RequestHandler + Send>
     where
         Self: 'static + Sized,
     {
         box self
     }
 }
 
-#[derive(Debug)]
-pub struct ReqContext {
-    pub context: kvrpcpb::Context,
-    pub table_scan: bool, // Whether is a table scan request.
-    pub txn_start_ts: u64,
-    pub start_time: Instant,
-    pub deadline: Instant,
+#[derive(Debug, Clone, Copy)]
+pub struct Deadline {
+    /// Used to construct the Error when deadline exceeded
+    tag: &'static str,
+
+    start_time: Instant,
+    deadline: Instant,
 }
 
-impl ReqContext {
-    pub fn new(ctx: &kvrpcpb::Context, txn_start_ts: u64, table_scan: bool) -> ReqContext {
+impl Deadline {
+    /// Initialize a deadline that counting from current.
+    pub fn from_now(tag: &'static str, after_duration: Duration) -> Self {
         let start_time = Instant::now_coarse();
-        ReqContext {
-            context: ctx.clone(),
-            table_scan,
-            txn_start_ts,
+        let deadline = start_time + after_duration;
+        Self {
+            tag,
             start_time,
-            deadline: start_time,
+            deadline,
         }
     }
 
-    pub fn set_max_handle_duration(&mut self, request_max_handle_duration: Duration) {
-        self.deadline = self.start_time + request_max_handle_duration;
-    }
-
-    #[inline]
-    pub fn get_scan_tag(&self) -> &'static str {
-        if self.table_scan {
-            "select"
-        } else {
-            "index"
-        }
-    }
-
-    pub fn check_if_outdated(&self) -> Result<()> {
+    /// Returns error if the deadline is exceeded.
+    pub fn check_if_exceeded(&self) -> Result<()> {
         let now = Instant::now_coarse();
         if self.deadline <= now {
             let elapsed = now.duration_since(self.start_time);
-            return Err(Error::Outdated(elapsed, self.get_scan_tag()));
+            return Err(Error::Outdated(elapsed, self.tag));
         }
         Ok(())
     }
 }
 
-pub use self::dag::{ScanOn, Scanner};
-pub use self::endpoint::{
-    Host as EndPointHost, RequestTask, Task as EndPointTask, DEFAULT_REQUEST_MAX_HANDLE_SECS,
-};
+/// Denotes for a function that builds a `RequestHandler`.
+/// Due to rust-lang#23856, we have to make it a type alias of `Box<..>`.
+type RequestHandlerBuilder<Snap> =
+    Box<dyn for<'a> FnBox(Snap, &'a ReqContext) -> Result<Box<dyn RequestHandler + Send>> + Send>;
```

Comment:
em, why add `dyn for` here?

Reply:
dyn is the new (recommended) syntax for specifying trait objects. If you don't write it, it will be implicitly written. This just make it clear. See https://github.com/rust-lang/rfcs/blob/master/text/2113-dyn-trait-syntax.md

Reply:
`for<'a>` is a HRTB here, see https://doc.rust-lang.org/nomicon/hrtb.html

Reply:
This is some seriously advanced lifetime usage. :+1:

Reply:
@Hoverbear Yes! And very useful here, otherwise you cannot write its type. (and there are compiler bugs 🤣